File: cousins_bfs.py
#time-n, space-h
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        q=deque()
        q.append(root)
        cnt=0
        while q:
            size=len(q)
            print(size)
            cnt=0
            for i in range(size):
                curr=q.popleft()
                if curr.left and curr.right:
                    if (curr.left.val==x and curr.right.val==y) or (curr.left.val==y and curr.right.val==x):
                        print("x")
                        return False
                
               
                if curr.left:
                    if curr.left.val==x or curr.left.val==y:
                        cnt+=1
                    q.append(curr.left)
                if curr.right:
                    if curr.right.val==x or curr.right.val==y:
                        cnt+=1
                    q.append(curr.right)
            print(cnt)
            if cnt==2:
                return True
        return False
                    
                
        

File: CousinsBFS.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only
class Solution {
    TreeNode xParent;
    TreeNode yParent;
    public boolean isCousins(TreeNode root, int x, int y) {
        if(root == null){
            return false;
        }
        if(x == y){
            return false;
        }
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        
        while(!queue.isEmpty()){
            int size = queue.size();
            boolean xFound = false;
            boolean yFound = false;
            for(int i = 0; i < size; i++){
                TreeNode current = queue.poll();
                if(current.left != null){
                    if(current.left.val == x){
                        xParent = current;
                        xFound = true;
                    }
                    if(current.left.val == y){
                        yParent = current;
                        yFound = true;
                    }
                    
                    queue.add(current.left);
                }
                if(current.right != null){
                    if(current.right.val == x){
                        xParent = current;
                        xFound = true;
                    }
                    if(current.right.val == y){
                        yParent = current;
                        yFound = true;
                    }
                    queue.add(current.right);
                }
            }
            
            if(xFound && yFound){
                if(xParent != yParent){
                    return true;
                }else{
                    return false;
                }                
            }
            
            if(xFound || yFound){
                return false;
            }
        }
        return false;        
    }
}